src/mongoose_traffic.erl

```diff
@@ -0,0 +1,127 @@
+-module(mongoose_traffic).
+
+%% The most simple use case possible is:
+%% - add {mongoose_debug, []} to modules in mongooseim.cfg
+%% - from erlang shell, run recon_trace:calls([{mongoose_debug, traffic, '_'}], 100, [{scope, local}]).
+%% - watch all the traffic coming in and out
+
+-behaviour(gen_mod).
+-behaviour(gen_server).
+
+-include("mongoose.hrl").
+-include("jlib.hrl").
+
+%% gen_mod API
+-export([start/2, stop/1]).
+%% hook handler
+-export([trace_traffic/3]).
+%% gen_server
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2]).
+%% cowboy handler for serving main page
+-export([init/2]).
+
+-define(SERVER, ?MODULE).
+
+
+start(HostType, _Opts) ->
+    gen_hook:add_handlers(hooks(HostType)),
+    case whereis(?MODULE) of
+        undefined ->
+            Traffic = {mongoose_traffic,
+                         {gen_server, start_link, [?MODULE, [], []]},
+                         permanent, 1000, supervisor, [?MODULE]},
+            % there has to be another layer
+            % channel will set up its own traces, someone has to watch and distribute stanzas
+            ejabberd_sup:start_child(Traffic);
+        _ ->
+            ok
+    end,
+    ok.
+
+stop(Host) ->
+    gen_hook:delete_handlers(hooks(Host)),
+    supervisor:terminate_child(ejabberd_sup, ?MODULE),
+    supervisor:delete_child(ejabberd_sup, ?MODULE),
+    ok.
+
+hooks(_HostType) ->
+    [{c2s_debug, global, fun ?MODULE:trace_traffic/3, #{}, 50}].
+
+
+trace_traffic(Acc, #{arg := {out, From, El}}, _) ->
+    traffic(out, From, El),
+    {ok, Acc};
+trace_traffic(Acc, #{arg := {in, El}}, _) ->
+    To = case get_jid(stanza, El) of
```

Comment:
This might be simplified to `get_jid(El, Acc)` with clauses that match on mongoose_acc first, then try to retrieve JID from `#xmlel` and return undefined if it's not there.